### PR TITLE
[SYCL-MLIR][LoopInternalization] Remove codegen dependency on `nd_item`

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Change code generation of `LoopInternalization` transformation to not rely on `nd_item`. Extend the transformation to allow kernel with input `sycl::item` on top of `sycl::nd_item`.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 1]
```diff
@@ -166,11 +167,14 @@ bool isCandidateKernel(gpu::GPUFuncOp kernel) {
...
+/// argument.
 bool isCandidateFunction(FunctionOpInterface func) {
-  // TODO: construct nd_item when not passed in.
-  if (func.getNumArguments() == 0 ||
-      !sycl::isPtrOf<sycl::NdItemType>(func.getArgumentTypes().back()))
+  if (func.getNumArguments() == 0)
+    return false;
+  Type lastArgTy = func.getArgumentTypes().back();
+  if (!sycl::isPtrOf<sycl::NdItemType>(lastArgTy) &&
+      !sycl::isPtrOf<sycl::ItemType>(lastArgTy))
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:177]: Can we get a `sycl.id` here too?

Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:177]: Likely we can, but the memory access analysis doesn't support that yet. How about we extent it further in another PR?

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 2]
```diff
@@ -535,35 +544,43 @@ Value castToIndex(Value val) {
...
-  return builder.create<sycl::SYCLNDItemGetLocalRangeOp>(
-      builder.getUnknownLoc(), builder.getI64Type(), ndItem,
-      builder.create<arith::ConstantIntOp>(builder.getUnknownLoc(), dim,
-                                           builder.getI32Type()));
+void WorkGroupSize::populateWorkGroupSize(unsigned numDims, OpBuilder builder) {
+  const auto arrayType = builder.getType<sycl::ArrayType>(
+      numDims, MemRefType::get(numDims, builder.getIndexType()));
+  const auto rangeTy = builder.getType<sycl::RangeType>(numDims, arrayType);
+  Location loc = builder.getUnknownLoc();
+  auto wgSize = builder.create<sycl::SYCLWorkGroupSizeOp>(loc, rangeTy);
```
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:574]: Can you add a utility function `createWorkgroupSize` in the SYCL `Utils`  similar to what we already have (e.g. createSYCLRangeGetOp).

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 3]
```diff
@@ -53,7 +53,8 @@ namespace {
 class WorkGroupSize {
 public:
   using ElemTy = std::variant<Value, unsigned>;
-  WorkGroupSize(Value ndItem, const sycl::ReqdWorkGroupSize &reqdWorkGroupSize,
+  WorkGroupSize(unsigned numDims,
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:59]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp:179]: added.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 4]
```diff
@@ -535,35 +544,43 @@ Value castToIndex(Value val) {
...
-  auto ndItemTy = cast<sycl::NdItemType>(
-      cast<MemRefType>(ndItem.getType()).getElementType());
-  const unsigned numDims = ndItemTy.getDimension();
   assert((reqdWorkGroupSize.empty() || reqdWorkGroupSize.size() == numDims) &&
          "Expecting reqdWorkGroupSize to have same number of elements as "
-         "nd_item dimension");
-  for (unsigned dim = 0; dim < numDims; ++dim) {
-    /// Use constants provided by reqd_work_group_size if given (non-empty).
-    if (!reqdWorkGroupSize.empty())
+         "numDims");
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp [hunk 5]
```diff
@@ -486,6 +481,20 @@ SmallVector<Value> getIndexes(sycl::SYCLAccessorSubscriptOp accSub,
   return indexes;
 }
 
+/// Return the grid dimension for \p func (e.g the dimension of the
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp]: done.

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h [hunk 6]
```diff
@@ -40,6 +45,11 @@ SYCLAccessorSubscriptOp createSYCLAccessorSubscriptOp(AccessorPtrValue accessor,
                                                       OpBuilder builder,
                                                       Location loc);
 
+/// Populate \p wgSizes with workgroup size per dimensionality, given the rank
+/// \p numDims.
+void populateWorkGroupSize(SmallVectorImpl<Value> &wgSizes, unsigned numDims,
```
Comment (etiotto) [inline mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:50]: Why here? This is a specialized function that make sense to have only in the transformation that uses `LoopInternalization`.

File: mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp [hunk 7]
```diff
@@ -49,3 +58,21 @@ sycl::createSYCLAccessorSubscriptOp(AccessorPtrValue accessor,
...
   return builder.create<SYCLAccessorSubscriptOp>(loc, MT, accessor, id);
 }
+
+void sycl::populateWorkGroupSize(SmallVectorImpl<Value> &wgSizes,
+                                 unsigned numDims, OpBuilder builder) {
+  const auto arrayType = builder.getType<sycl::ArrayType>(
+      numDims, MemRefType::get(numDims, builder.getIndexType()));
+  const auto rangeTy = builder.getType<sycl::RangeType>(numDims, arrayType);
+  Location loc = builder.getUnknownLoc();
+  auto wgSize = builder.create<sycl::SYCLWorkGroupSizeOp>(loc, rangeTy);
```
Comment (etiotto) [inline mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp:68]: As I mentioned in my previous code review please create a function here called `createWorkGroupSize` and move back this function in LoopInternalization.

Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp:68]: I thought you want `populateWorkGroupSize` to be moved to Utils.

Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp]: Added.

Comment (etiotto) [reply mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp]: My opinion is that `populateWorkGroupSize` is a specialized function that is useful to the LoopInternalization pass. Not a generic utility function that would be useful to the SYCL dialect. 

Perhaps other ppl will have a different opinion, but that is how I feel.

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h (hunk 6 above)
Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:54]: It is useful to the transformation `LoopInternalization` but I do not think it belongs here.

Comment (whitneywhtsang) [reply mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:54]: Ettore's other comment: https://github.com/intel/llvm/pull/9894#discussion_r1231354471:

> My opinion is that `populateWorkGroupSize` is a specialized function that is useful to the LoopInternalization pass. Not a generic utility function that would be useful to the SYCL dialect. 
>
> Perhaps other ppl will have a different opinion, but that is how I feel.

File: mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp (hunk 7 above)
Comment (etiotto) [reply mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp]: No it is up to the person that has a question to resolve a thread. Not to the author of the PR.

Comment (whitneywhtsang) [reply mlir-sycl/lib/Dialect/SYCL/Utils/Utils.cpp]: ok, let's keep the two threads then.

File: mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h (hunk 6 above)
Comment (etiotto) [reply mlir-sycl/include/mlir/Dialect/SYCL/Utils/Utils.h:54]: I am going to let this PR merge in, given that the contentious matter is not critical. I will post a separate PR to move this function and `constructSYCLID` to LoopInternalization.

**Review by etiotto — APPROVED** [review]

**Pull request merged**